### PR TITLE
add GPT boot partition support via boot-gpt-switch slot type

### PR DESCRIPTION
**Pull request by jluebbe**
This adds support for atomic update of bootloaders in the first GPT partitions (such as the ESP), similar to what we already support for MBR (via boot-mbr-switch).

The PR also adds support for testing of both boot-switch methods in qemu on a loopback device.

### CODE REVIEW

File: include/gpt.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+#include <glib.h>
+
+#include "update_handler.h"
+
+/**
+ * Get the address and size of the inactive boot partition
+ * if a partition exists in the defined region.
+ *
+ * @param device dev path (/dev/mmcblkX)
```
Comment (OnkelUlla) [inline include/gpt.h:11]: In the forthcoming commit 94a310115eea you explicitely used `/dev/sda` in the corresponding section of `system.conf`. So perhaps use `/dev/sdX` instead of `/dev/mmcblkX` here as well? The same applies to line 29 in this file.

File: src/gpt.c [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+		goto out;
+	}
+
+	number_of_sectors = fdisk_get_nsectors(cxt);
+	if ((region_start + region_size) / sector_size >= number_of_sectors) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region configuration is bigger than device");
+		res = FALSE;
+		goto out;
+	}
```
Comment (OnkelUlla) [inline src/gpt.c:134]: Why not additionally sanity-check `region_start` for its own in the same way as `region_start + region_size`?

File: src/gpt.c [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+				"Region size %"G_GINT64_MODIFIER "d is not aligned to the double grain-size %"G_GINT64_MODIFIER "d",
+				region_size, 2 * grain_size);
+		goto out;
+	}
+
+	number_of_sectors = fdisk_get_nsectors(cxt);
+	if ((region_start + region_size) / sector_size >= number_of_sectors) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Region configuration is bigger than device");
+		res = FALSE;
```
Comment (OnkelUlla) [inline src/gpt.c:132]: Why do you explicitely assign that to `res` here? You don't do that in the rest of this code.
```suggestion
```

File: src/gpt.c [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+		goto out;
+	}
+
+	sector_size = fdisk_get_sector_size(cxt);
```
Comment (OnkelUlla) [inline src/gpt.c:136]: This has already been done in line 106.
```suggestion
```

File: src/gpt.c [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+			found = TRUE;
+			continue;
+		}
+
+		p_start_sector = fdisk_partition_get_start(pa);
+		p_end_sector = fdisk_partition_get_end(pa);
+
+		if (region_start_sector <= p_end_sector && p_start_sector <= region_end_sector) {
+			g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+					"Region (sectors 0x%"G_GINT64_MODIFIER "x - 0x%"G_GINT64_MODIFIER "x) overlaps"
```
Comment (OnkelUlla) [inline src/gpt.c:167]: There is a single space missing:
```suggestion
					"Region (sectors 0x%"G_GINT64_MODIFIER "x - 0x%"G_GINT64_MODIFIER "x) overlaps "
```

File: src/gpt.c [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+			goto out_table;
+		}
+	}
+
+	if (!found) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"No boot partition found in entry %d",
+				BOOT_PARTITION_ENTRY);
+		goto out_table;
+	}
```
Comment (OnkelUlla) [inline src/gpt.c:181]: The check for the existence of a boot partition is done in lines +226 to +231 of `r_gpt_switch_get_inactive_partition()` and lines +307 to +312 of `r_gpt_switch_set_boot_partition()` as well. Why have it here in a different form? Are there cases that we are not going to catch if we delete this check here?

File: src/gpt.c [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+
+	sector_size = fdisk_get_sector_size(cxt);
+	fdisk_reset_partition(pa); /* only change the location */
+	fdisk_partition_set_start(pa, partition->start / sector_size);
+	fdisk_partition_set_size(pa, partition->size / sector_size);
+
+	if (fdisk_set_partition(cxt, BOOT_PARTITION_ENTRY, pa) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not update boot partition");
+		goto out_deassign;
```
Comment (OnkelUlla) [inline src/gpt.c:322]: Here the partition `pa` is already a valid reference so it has to be unreferenced:
```suggestion
		goto out_unref_part;
```

File: src/gpt.c [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+	}
+
+	if (fdisk_verify_disklabel(cxt) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"New GPT failed to verify");
+		goto out_unref_part;
+	}
+	if (fdisk_write_disklabel(cxt) != 0) {
```
Comment (OnkelUlla) [inline src/gpt.c:330]: Does `fdisk_write_disklabel()` give the required guarantee of being atomic? I will refer to that in your documentation update.

File: src/update_handler.c [hunk 9]
```diff
@@ -1622,6 +1738,11 @@ RaucUpdatePair updatepairs[] = {
 	{"*.vfat", "boot-mbr-switch", img_to_boot_mbr_switch_handler},
 	{"*.img", "boot-mbr-switch", img_to_boot_mbr_switch_handler},
 	{"*", "boot-mbr-switch", NULL},
+#if ENABLE_GPT == 1
+	{"*.vfat", "boot-gpt-switch", img_to_boot_mbr_switch_handler},
```
Comment (OnkelUlla) [inline src/update_handler.c:1742]: Here is a copy-n-paste leftover:
```suggestion
	{"*.vfat", "boot-gpt-switch", img_to_boot_gpt_switch_handler},
```

File: src/update_handler.c [hunk 10]
```diff
@@ -1360,6 +1361,121 @@ static gboolean img_to_boot_mbr_switch_handler(RaucImage *image, RaucSlot *dest_
...
+	g_message("Copying image to inactive boot partition %d on %s",
+			inactive_part, dest_slot->device);
+
+	res = copy_raw_image(image, outstream, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	g_message("Setting GPT to switch boot partitions");
```
Comment (OnkelUlla) [inline src/update_handler.c:1455]: You explicitely switch from `partitions` to `partition` in the documentation update. Should it become singular here as well?

File: docs/advanced.rst [hunk 11]
```diff
@@ -740,7 +740,7 @@ A ``system.conf`` could look like this:
 
 .. _sec-mbr-partition:
 
-Update Boot Partitions in MBR
+Update Boot Partition in MBR
```
Comment (OnkelUlla) [inline docs/advanced.rst:743]: This is the line that I refered to where you explicitely switch to singular.

File: docs/advanced.rst [hunk 12]
```diff
@@ -740,7 +740,7 @@ A ``system.conf`` could look like this:
...
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (OnkelUlla) [inline docs/advanced.rst:744]: If removing the plural 's' in the line above then one '~' should be removed here as well.

File: docs/advanced.rst [hunk 13]
```diff
@@ -799,6 +800,53 @@ A ``system.conf`` could look like this:
   region-start=164352
   region-size=66M
 
+.. _sec-gpt-partition:
+
+Update Boot Partitions in GPT
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (OnkelUlla) [inline docs/advanced.rst:806]: Nitpick: singular or plural?
```suggestion
Update Boot Partition in GPT
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: docs/advanced.rst [hunk 14]
```diff
@@ -799,6 +800,53 @@ A ``system.conf`` could look like this:
...
+
+Systems booting via UEFI have a special partition, called the *EFI system
+partition (ESP)*, which contains the bootloader to be started by the UEFI
+firmware.
+Also, some newer ARM SoCs support loading the bootloader directly from a GPT
+partition.
+
+To allow atomic updates of these partitions, RAUC supports changing the GPT to
```
Comment (OnkelUlla) [inline docs/advanced.rst:814]: Is `fdisk_write_disklabel()` really guaranteed to be atomic?

File: docs/advanced.rst [hunk 15]
```diff
@@ -799,6 +800,53 @@ A ``system.conf`` could look like this:
...
+The device to be specified is expected to be the underlying block device.
+The boot partitions are derived by the definition of the values ``region-start``
+and ``region-size``.
+Both values have to be set in integer decimal bytes and can be post-fixed with
+K/M/G/T.
+
+To ensure that the resulting GPT entries are well aligned, the region start must
+be a multiple of the *grain* value (as used by ``sfdisk``), which is 1MB by
+default.
+Accordingly, the region size must be a aligned to twice the *grain* value (to
```
Comment (OnkelUlla) [inline docs/advanced.rst:832]: Typo:
```suggestion
Accordingly, the region size must be aligned to twice the *grain* value (to
```

**Comment by OnkelUlla** [comment]
Good work, Jan!

File: src/gpt.c [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+	}
+
+	sector_size = fdisk_get_sector_size(cxt);
+	region_start_sector = region_start / sector_size;
+	region_end_sector = region_start_sector + region_size / sector_size - 1;
+
+	if (fdisk_get_partitions(cxt, &tb)) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Failed to get partitions");
+		return FALSE;
```
Comment (OnkelUlla) [inline src/gpt.c:143]: To stay consistent presumably this should be:
```suggestion
		goto out;
```

File: src/gpt.c (hunk 4 above)
Comment (jluebbe) [reply src/gpt.c:136]: Line 106 is for the sector size. Nevertheless, I've moved it to the related getter calls.

File: src/gpt.c (hunk 8 above)
Comment (jluebbe) [reply src/gpt.c:330]: Yes. It follows a specific order of writes to make sure it has always a working copy.

File: src/gpt.c (hunk 6 above)
Comment (jluebbe) [reply src/gpt.c:181]: We reload the partition table, so I wanted to be extra sure.

Comment (OnkelUlla) [reply src/gpt.c:185]: Yes, that sounds reasonable.

File: src/gpt.c (hunk 8 above)
Comment (OnkelUlla) [reply src/gpt.c:338]: Okay, thanks for the explanation. I just wanted to be sure, but did not look into the internals of `fdisk_write_disklabel()`.

File: src/gpt.c [hunk 17]
```diff
@@ -0,0 +1,354 @@
...
+		goto out_unref_part;
+	}
+
+	if (fdisk_verify_disklabel(cxt) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"New GPT failed to verify");
+		goto out_unref_part;
+	}
+	/* As we made sure that we have two copies above, we are sure to always
+	 * have a least one valid copy at any point during the update. */
```
Comment (OnkelUlla) [inline src/gpt.c:337]: In this good new comment a small typo slipped in:
```suggestion
	 * have at least one valid copy at any point during the update. */
```

File: test/common.c [hunk 18]
```diff
@@ -217,6 +217,7 @@ gboolean test_make_filesystem(const gchar *dirname, const gchar *filename)
 			&error,
 			"/sbin/mkfs.ext4",
 			"-F",
+			"-I256",
```
Comment (ejoerns) [inline test/common.c:220]: This is unrelated and maybe of increasing significance for the users of RAUC. Maybe it would be worth making a separate PR for documentation purposes?

File: docs/advanced.rst [hunk 19]
```diff
@@ -799,6 +800,53 @@ A ``system.conf`` could look like this:
...
+A ``system.conf`` section could look like this:
+
+.. code-block:: cfg
+
+  [slot.esp.0]
+  device=/dev/sda
+  type=boot-gpt-switch
+  region-start=1M
+  region-size=32M
+
```
Comment (ejoerns) [inline docs/advanced.rst:847]: nit: at least one excess empty line here

File: src/gpt.c [hunk 20]
```diff
@@ -0,0 +1,354 @@
...
+	}
+	if (fdisk_get_alignment_offset(cxt) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Non-zero aligment offset (%ld) is not supported",
+				fdisk_get_alignment_offset(cxt));
+		goto out;
+	}
+
+	if (grain_size < sector_size)
+		g_error("%s: libfdisk reported grain size is less than sector size\n", G_STRLOC);
```
[the remark this answers is not in the document]
Comment (jluebbe) [reply src/gpt.c:82]: I've removed the `\n` and moved it up. If this happens, something is very wrong, so a coredump is appropriate.

File: test/common.c (hunk 18 above)
Comment (jluebbe) [reply test/common.c:220]: Yes, will drop it from this PR.

**Comment by ejoerns** [comment]
Rebased onto master to resolve conflict in `CHANGES`.
